Fix Kernel normalization

larrybradley opened the pull request:
This PR fixes an issue with kernel normalization.  Currently the normalization factor is defined when the kernel is initialized and never updated.  Consequently, repeatedly running `.normalize()` will continue to apply the original normalization factor.  In this PR the normalization factor is recomputed each time `.normalize()` is called.

Here's an example showing the current bug:
```
import numpy as np
from astropy.convolution import CustomKernel

kernel = CustomKernel(np.ones(5))
kernel.normalize()
print kernel.array
[ 0.2  0.2  0.2  0.2  0.2]

kernel.normalize()
print kernel.array
[ 0.04  0.04  0.04  0.04  0.04]
```

larrybradley commented:
`coveralls` failed for a -0.0% decreased coverage(!), so I added several kernel tests (and fixed some logic issues that those tests uncovered) and did some general clean up in the `convolution` tests.

larrybradley commented:
Can someone please restart `travis` test 9304.9 (transient failure)?  Thanks.

larrybradley commented:
Anyone care to review this?

larrybradley commented:
@adonath?

adonath commented:
I'm happy to review this next week!

adonath commented on astropy/convolution/core.py:
```diff
@@ -88,42 +85,66 @@ def center(self):
+        else:
+            raise ValueError("invalid mode, must be 'integral' or 'peak'")
+        return self._normalization
+
```
Is it necessary to introduce the new function `calculate_normalization`? As far as I can tell it does the same thing as the former `normalize` function (except that it fixes the bug, with repeated normalization, of course...). Couldn't this be added to the `normalize` function? If not I would suggest to make the function private, otherwise there is an overlap with the `normalization` property (both return `self.normalization`...).

larrybradley replied:
Thanks for reviewing this PR.

They are separate because `Kernel.__init__()` calls `calculate_normalization` to define `self._normalization` (but this does *not* normalize the kernel).  `self._normalization` is currently needed in `Kernel.__init__()` by `Gaussian1DKernel` and `Gaussian2DKernel` to define `self._truncation`.

The `normalize` method calls `calculate_normalization` and then also normalizes the kernel.

So, I think the options are

 1. Leave this as is.

 2. Combine `calculate_normalization` and `normalize` into a single method, remove `calculate_normalization` from `Kernel.__init__`, and in `Gaussian1DKernel` and `Gaussian2DKernel` explicitly call `self.normalize()` before defining `self._truncation`.

 3. Combine `calculate_normalization` and `normalize` into a single method, call `self.normalize()` in `Kernel.__init__`, and then remove all the separate `self.normalize()` calls in the kernel definitions (in `kernels.py`).  This would automatically normalize all kernels at initialization.  Is that desired?

Any preference?

larrybradley replied:
I guess I'm leaning towards Option 2.  But if the kernels should always automatically be normalized, then Option 3.

adonath replied:
I had to think about this a little bit again (and remember the intention of this normalization functions...). I think the idea behind the `self._normalization` attribute was, that it should contain the value you have to multiply the kernels `self._array` with, so that it sums to `1`. So `self._normalization` corresponds to `1. / self._array.sum()`, regardless whether the normalization is `peak` or `integral`. The reason for this is, that the astropy convolution algorithms require a normalized kernel (i.e. a kernel that sums to 1). In that sense the bug could be just fixed by setting `self._normalization = 1.` in the  `self.normalize` function, when using the `integral` mode. Is that correct?

larrybradley replied:
Thanks.  It wasn't at all clear to me that `self._normalization` should *always* be `1. / self._array_sum()`.  Actually, I don't think `self._normalization` is really needed at all, except for defining the `_truncation` in the `Gaussian` kernels.  The user should always call `self.normalize()` to normalize the kernel.

`astropy.convolution.convolve` actually doesn't *require* a normalized kernel, only if you want to conserve flux, in which case one should always call `self.normalize()` before `convolve` (not use `self._normalization`).

One problem with setting `self._normalization` to 1 after normalization (instead of recalculating it), is that if one reassigns `kernel.array` then `self._normalization` will be wrong.

adonath commented:
That was indeed a bug! Thanks for fixing this and adding the tests. Apart from the one comment I've left, it looks fine to me!

adonath commented on astropy/convolution/core.py:
```diff
@@ -88,42 +85,66 @@ def center(self):
+            else:
+                self._normalization = 1. / self._array.sum()
+        elif mode == 'peak':
+            self._normalization = 1. / self._array.max()
```
I think `self._normalization` should be set to `1. / self._array.sum()` regardless whether the mode is `peak` or `integral` (see comment below).

adonath commented on astropy/convolution/core.py:
```diff
@@ -88,42 +85,66 @@ def center(self):
+                          'large, > {0}.'.format(MAX_NORMALIZATION),
                           AstropyUserWarning)
-        if mode == 'integral':
-            self._array *= self._normalization
```
I think adding `self._normalization = 1` or `self._normalization = 1. / self._array.sum()` would fix the bug

adonath commented:
I thought about this again and I've left a few additional comments. I hope they make sense...

larrybradley commented:
I've updated this PR with the following changes:

- made `Kernel.array` always be a `numpy.ndarray`
- combined `calculate_normalization` and `normalize`
- removed defining `_normalization` in `Kernel.__init__` (it was
  needed only for `Gaussian` kernels).
- update the `_truncation` calculation when defining `Gaussian` kernels (see above)
- removed `_normalization` property
- added `self._kernel_sum`, which is calculated after normalization, but it should never really be needed

adonath commented:
Thanks for making the changes. Now it looks fine to me!

embray commented:
There still seems like there's a lot of boilerplate going on in the `__init__`s for individual Kernel classes that could be done away with, but this is still a big improvement :+1:
